### ISSUE TITLE
scatter_matrix formatting considered harmful

### ISSUE THREAD

**Issue by jbednar** [holoviz/hvplot#1285]
#### ALL software version info
Python              :  3.11.5 (main, Sep 11 2023, 08:31:25) [Clang 14.0.6 ]
Operating system    :  macOS-14.0-arm64-arm-64bit
Panel comms         :  default

holoviews           :  1.18.2
bokeh               :  3.3.4
colorcet            :  3.0.1
dask                :  2023.6.0
datashader          :  0.16.0
geoviews            :  1.11.0
hvplot              :  0.9.0
IPython             :  8.15.0
jupyterlab          :  4.0.11
matplotlib          :  3.8.0
notebook            :  7.0.6
numba               :  0.58.0
numpy               :  1.24.3
pandas              :  2.1.1
panel               :  1.3.1
param               :  2.0.1
pillow              :  10.0.1
pyarrow             :  11.0.0
pyviz_comms         :  2.3.0
scipy               :  1.11.3
spatialpandas       :  0.4.9
xarray              :  2023.6.0

#### Description of expected behavior and the observed behavior

I'd expect `scatter_matrix` to be formatted reasonably: subplots all lined up, axis labels readable, text not overlapping, and a single Bokeh toolbar for the entire figure. That's not what's happening:

```python
import pandas as pd
import hvplot.pandas
from hvplot import scatter_matrix

url = 'https://raw.githubusercontent.com/shoukewei/data/main/data-pydm/gdp_top_six_economies.csv'
df = pd.read_csv(url)
scatter_matrix(df, alpha=0.5, width=600, height=600, xrotation=0)
```
<img width="786" alt="image" src="https://github.com/holoviz/hvplot/assets/1695496/8b0486b1-f249-47fc-896d-c2b54ba8fdd9">

```python
scatter_matrix(df, alpha=0.5, width=600, height=600, xrotation=90)
```
<img width="777" alt="image" src="https://github.com/holoviz/hvplot/assets/1695496/d92cf4e8-355e-4250-b6bf-572bf29b9e3f">

```python

```

**Comment by droumis** [comment]
Just adding that the spacing issue seems unrelated to the toolbar:

`scatter_matrix(df, alpha=0.5, width=600, height=600, xrotation=0).opts(toolbar='above')`

![image](https://github.com/holoviz/hvplot/assets/6613202/8fc8c775-2785-493b-b14a-67a38e71dc01)

**Comment by jbednar** [comment]
> Just adding that the spacing issue seems unrelated to the toolbar:

Yes, I think the spacing issues have been there for some time, while the toolbar issue is relatively recent, but I haven't tried to do a git bisect to pin that down.

**Comment by maximlt** [comment]
Quick feedback:

- Toolbar issue: possibly introduced in https://github.com/holoviz/holoviews/pull/5873
- Subplots not aligned: happened sometime around the Bokeh 3 transition (not sure it's a Bokeh 3 problem either)


https://holoviews.org/reference/containers/bokeh/GridSpace.html
![image](https://github.com/holoviz/hvplot/assets/35924738/851814a6-31bc-475f-9e69-75e9686c9128)

There may be no hvPlot issue at all.

**Comment by jbednar** [comment]
Thanks. I've opened https://github.com/holoviz/holoviews/issues/6126 for the toolbar issue, and @mattpap is looking at it from the Bokeh side.

**Comment by mattpap** [comment]
Bad plot alignment is caused by fixed frame sizing (`Plot.frame_{width,height,align}`), which works reliably only for single plots and doesn't work well in all other cases (see e.g. issue https://github.com/bokeh/bokeh/issues/13225). I suppose it's time to implement this properly.

**Comment by maximlt** [comment]
From the initial list of issues:

1. [ ] subplots all lined up: Matteusz indicated this is a Bokeh issue
2. [x] a single Bokeh toolbar for the entire figure: HoloViews issue fixed in https://github.com/holoviz/holoviews/pull/6127
3. [ ] axis labels readable, text not overlapping

That leaves us with 3). The default Bokeh formatter is the `BasicTickFormatter`:

<img width="683" alt="image" src="https://github.com/holoviz/hvplot/assets/35924738/131bf029-5684-4bcd-bbfd-95f02cdb2a12">
<img width="396" alt="image" src="https://github.com/holoviz/hvplot/assets/35924738/34d0db3e-adb1-43b8-9c70-73174c7446ef">

Comparing that to the default of plotly express:
<img width="552" alt="image" src="https://github.com/holoviz/hvplot/assets/35924738/6f47fe1b-8acc-4f69-ad1d-4a4461bbf2b9">

We can get a similar behavior defining a `NumericalTickFormatter`:
<img width="656" alt="image" src="https://github.com/holoviz/hvplot/assets/35924738/010567d7-3d5b-4cac-9f9b-d6a65991fc2c">

However, it also has its limits:
<img width="649" alt="image" src="https://github.com/holoviz/hvplot/assets/35924738/dbbef5ff-9498-4ea2-8363-cca7d5abc5b5">

Certainly, we could better document `xformatter/yformatter`. But should we also consider defaulting to a more user-friendly formatter?

**Comment by jbednar** [comment]
Defaulting to a more usable formatter sounds like a great idea. @mattpap , any idea why the tick formatter didn't decide to drop the intermediate tick marks? Here I'd be hoping to get one label on the left of the x axis, and one on the right:

<img width="68" alt="image" src="https://github.com/holoviz/hvplot/assets/1695496/91d0532f-2722-4975-87e9-84ddefc653f7">

**Comment by mattpap** [comment]
> any idea why the tick formatter didn't decide to drop the intermediate tick marks?

This is handled setting `Axis.major_label_policy = NoOverlap()`. When this was implemented the default (`AllLabels`) was left for backwards compatibility. Tickers and tick formatters have no access to the screen space, so they can't make any adjustments based on the positioning of labels.

**Comment by jbednar** [comment]
Thanks! Ok, @maxime, can you try out the new option with NumericalTickFormatter? For hvPlot I strongly favor improving the user experience over preserving previous defaults.
